addons/l10n_id/tests/test_qris.py

```diff
@@ -0,0 +1,199 @@
+            'company_id': cls.company_data['company'].id,
+            'invoice_line_ids': [Command.create({'quantity': 1, 'price_unit': 100})],
+        })._post()
+
+        cls.success_qris_get = {
+            "status": "success",
+            "data": {
+                "qris_content": "Test Content",
+                "qris_request_date": "2024-02-27 11:13:42",
+                "qris_invoiceid": "413255111",
+                "qris_nmid": "ID1020021181745"
+            }
+        }
+        cls.success_qris_get_second = {
+            "status": "success",
+            "data": {
+                "qris_content": "Test Content Second",
+                "qris_request_date": "2024-02-27 11:13:42",
+                "qris_invoiceid": "413255111",
+                "qris_nmid": "ID1020021181745"
+            }
+        }
+
+        cls.qris_status_fail = {
+            "status": "failed",
+            "data": {
+                "qris_status": "unpaid"
+            }
+        }
+
+        cls.qris_status_success = {
+            "status": "success",
+            "data": {
+                "qris_status": "paid",
+                "qris_payment_customername": "Zainal Arief",
+                "qris_payment_methodby": "Sakuku"
+            },
+            "qris_api_version_code": "2206091709"
+        }
+
+    @freeze_time("2024-02-27 04:15:00")
+    def test_qris_qr_code_generation(self):
+        """ QR-Code generation conditions:
+
+        - should only come from portal side
+        - only call the API when the QRIS transaction not found or lasted for > 30 minutes
+        - if transaction found and < 30 minutes, no API call needed
+        """
+        self.qris_qr_invoice.qr_code_method = 'id_qr'
+
+        with patch(
+            'odoo.addons.l10n_id.models.res_bank.ResBank._l10n_id_make_qris_request', return_value=self.success_qris_get
+        ) as patched:
+            # QR code shouldn't be generated without the context.
+            result = self.qris_qr_invoice._generate_qr_code()
+            self.assertIsNone(result)
+
+            # But of course, should be with it.
+            result = self.qris_qr_invoice.with_context({'is_online_qr': True})._generate_qr_code()
+            self.assertIsNotNone(result)
+
+            # Confirm that the QR was successfully registered on the invoice.
+            qr_details = self.qris_qr_invoice.l10n_id_qris_invoice_details
+            self.assertIsNotNone(qr_details)
+
+            qris_data = self.success_qris_get['data']
+            self.assertEqual(qr_details[0]['qris_invoice_id'], qris_data['qris_invoiceid'])
+            self.assertEqual(self.qris_qr_invoice.l10n_id_latest_qris_content, qris_data['qris_content'])
+            patched.assert_called_once()
+
+            # Check the API is not called again, as it should reuse the existing QR until it is expired
+            self.qris_qr_invoice.with_context({'is_online_qr': True})._generate_qr_code()
+            patched.assert_called_once()
```

Comment:
It's weird to have 2 `assert_called_once` on the same patch.
I would do the same patch twice, second time with `assert_not_called`, or at least put a comment